[docs] node to node protocol.

majestrate opened the pull request:
rough draft outline of lokinet's node to node protocol.

majestrate committed:
REWORDME: [docs] node to node protocol.

rough draft outline of lokinet's node to node protocol.

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,105 @@
+
+## streamed requests
+
+direct node to node request/response messages sent on a bidi quic stream.
+
+the stream starts by sending 2 bytes identifying the stream kind of which there are the following:
+
+* network data lookup
+* network data upsert
+* path build
+* path update
+* path stream
```
It is not clear to me from this description how requests and streams interact.

Is the idea here that there is one stream per request?  Or one stream per *type* of request?

majestrate replied:
the idea is that streams are for one type of request, but there is no
requirement that you only have one per type. this is especially true
with path stream type.

looking at this again i think it would be best to define which streams
would allow multiple instances per remote. e.g. a path stream
connection would have many instances to each remote, and per path
per remote, but for network data lookup/upsert you'd really only need
one per remote.

majestrate replied:
clarification added

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,105 @@
+
+### network data lookup stream
+
+used for fetching of:
```
Used how?  Right now this reads a bit like a wishlist, but to properly document the protocol, this page needs to evolve into something that completely describes the messages.  Ultimately, I want me or you or some other dev to be sit down with *only* this document and have enough information to produce an implementation that can interact with a remote Lokinet via the wire protocol.

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,105 @@
+* additional extra fields used in the future.
+
+
+(todo: i want to permit multiple kinds of pubkeys for forwards compat)
```
I don't understand this todo: I don't see how you could be forwards compatible with a type of key you don't understand how to use.  To the extent that future versions of the protocol evolve to add other keys, you would either need to do that in a backwards-compatible way (i.e. providing alternatives, but still supporting the old key type for backwards compat), or in a breaking protocol change.

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,105 @@
+onion routed persisting stream connection to the pivot router for sending anonymised requests destined for the pivot router.
+
+streams are created for an existing path by opening a "path stream" stream providing the path's tx/rx id and 24 bytes nonce.
+on accepting the stream creation the node will open a "path stream" to their upstream router but with a nonce that is deterministically mutated `new_nonce = nonce ^ hash(nonce)`, this mutation reduces stream corrilation across hops.
```
> `new_nonce = nonce ^ hash(nonce)`

Just a hash is enough; the extra XOR of the nonce adds nothing.

majestrate replied:
this operation needs to be commutative as it is done in both directions. when we get a nonce as the path builder from the pivot router we need to be able to do each operation in reverse order.

jagerman replied:
I'm not understanding.

If I compute:

    X = N xor hash(N)

then how do I get N from X?

majestrate replied:
that was my point, you can't.

jagerman replied:
So then I'm back to my original comment: `X = N xor hash(N)` offers nothing over just `X = hash(N)`

majestrate replied:
they both suffer from the same problem. it's a one way operation. when the sender of the original message is not the path creator we cannot obtain the original value of X because it's gone through a hash function. what we do right now with our existing scheme is `N = N ^ hash(path_shared_secret_symettric_key)`, note that the operation applied is `^`, and the `hash()` part of it is from the frame of reference of the next hop is such that they dont really care that a hash even happens. this is a tangent, i'm redoing this entire pr as i misunderstood the level of detail required.

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,105 @@
+streams are created for an existing path by opening a "path stream" stream providing the path's tx/rx id and 24 bytes nonce.
+on accepting the stream creation the node will open a "path stream" to their upstream router but with a nonce that is deterministically mutated `new_nonce = nonce ^ hash(nonce)`, this mutation reduces stream corrilation across hops.
+on rejecting the stream creation the node will shutdown the read end of the stream, sending a reply signed by our identity key and then shutting down the write end of the stream (which fully closes the stream).
+if the accepting node is the pivot route it will write. 
```
It will write what?

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,105 @@
+on rejecting the stream creation the node will shutdown the read end of the stream, sending a reply signed by our identity key and then shutting down the write end of the stream (which fully closes the stream).
+if the accepting node is the pivot route it will write. 
+
+a "request stream path result" message has a flag for the result, zero for success and non zero for failure.
```
What else does the message have?  What does it look like, how is it encoded, etc?

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,105 @@
+
+### path streams
+
+onion routed persisting stream connection to the pivot router for sending anonymised requests destined for the pivot router.
```
I don't really understand why this is restricted to pivot routers: most of the requests listed in "path stream requests" seems like they would be invoked on any old remote router, not exclusively a pivot router.

jagerman replied:
Maybe a term such as "target router" would make more sense?

majestrate committed:
clarification

majestrate committed:
ammendment to logic.

nonce = nonce ^ hash(nonce) is only able to work one way. this is why
we use nonce = nonce ^ hash(path_ephermal_secret) for mutation in the
existing onion routing parts. update this line to use that logic
instead.

majestrate commented:
@jagerman i fell down an analytical rabbit hole last week on this so i forget if this was meant to be a full on technical protocol spec or just a high level thing.

remind me which one this was so i can dig into this again and so that that can be tracked.

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,105 @@
+(todo: if we use a ring signature on the "request path stream result" message can do a pop and shift on the pubkeys in the ring so that the original requestor can know who rejected/accepted without the intermediate hops knowing who did. would we have to generate derived keys here?)
+(todo: if we use normal signatures here would we want to blind the signing keys some how? rotating blinded keys rachet style)
+
+after built we have a quic substream which will do xchacha20 stream cipher at each hop using the nonce provided in the request. this substream can be used to send consensus data requests to the pivot router.
```
It's not clear to me how this encryption mechanism will work:
- xchacha20 requires not just a nonce but also a key
- this is only describing how forward encryption works; where do the keys + nonces for the response stream encryption come from?
- how to deal with xchacha20's 64-byte block size?
- where does authentication happen (to detect/reject tampering)?
- can we recover from lost state?  (i.e. when we lose the IC counter for the stream)
- how and when do stream errors get detected and signaled?  (e.g. suppose intermediate router 2 has gone away).

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,105 @@
+
+(todo: finish this)
+
+### path streams
```
This section left me with a lot of questions and took me a considerable amount of time to try to get my head around.  I feel like there is a lot of omitted details here in terms of describing how this gets initiated, how intermediate sub-streams get established, how data moves along it, how it gets authenticated, and so on.  There are *pieces* of all of those in here, but I feel a bit like I'm reading a book with every other sentence removed and trying to guess what the big picture is.

I need a diagram or a detailed description or both of how these paths work conceptually before you start in on the weeds of path ids, nonces, and encryption, because the picture I have in my head right now is a guess more than anything.

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,105 @@
+
+## lossy traffic
+
+one shot datagrams with a 24 byte nonce and tx/rx path id are sent best effort on an established path with the nonce mutated deterministically and symettric encryption applied. 
```
Elaboration needed.

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,105 @@
+
+sent on an existing path to update metadata about the path.
+
+lets us close a path prematurely. could let us extend the lifetime of a path. could let us rewire an upstream router and path id at a hop ( this can be used to route around a bad snode without building a full new path ).
```
All of these capabilities, requests, responses, etc. needs to be fleshed out.

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,107 @@
+
+direct node to node request/response messages sent on a bidi quic stream.
+
+a stream carries 1 type of request/response messages. we maintain 1 stream per type to each remote node with the exception of path stream which requires multiple instances of itself.
```
I don't really understand the point of using one stream per message type; since it isn't one stream per request, we need a request/response framework on top of the stream at which point you might as well just put the request type into the request itself.

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,107 @@
+
+onion routed persisting stream connection to the pivot router for sending anonymised requests destined for the pivot router.
+
+streams are created for an existing path by opening a "path stream" stream providing the path's tx/rx id and 24 bytes nonce.
```
The topology is not clear.

Suppose client A talks directly to 2 entry routers (E₁, E₂); through each entry router she builds random paths to three "target" routers (T₁, T₂, T₃ through E₁; and T₄, T₅, T₆ through E₂) to which she wants to make path stream requests; we'll suppose paths have two intermediate routers (X₁ through X₁₂).

Thus, from A's point of view, there are 6 paths:

- A-E₁-X₁-X₂-T₁
- A-E₁-X₃-X₄-T₂
- A-E₁-X₅-X₆-T₃
- A-E₂-X₇-X₈-T₄
- A-E₂-X₉-X₁₀-T₅
- A-E₂-X₁₁-X₁₂-T₆

What are all of the corresponding QUIC streams on the network, and how much state is attached to each?

jagerman commented on docs/node2node.md:
```diff
@@ -0,0 +1,107 @@
+used for fetching of:
+
+* pubkey list with unlock times if they are applicable.
+* pubkey list delta given last hash of full list.
```
The purpose of these two endpoints is not obvious to me.  (Not saying there isn't one, just that it isn't obvious from just this description).

jagerman reviewed (commented):
This PR is a good start of the new protocol design.  There is, however, still considerable augmentation, elaboration, and details needed (both high-level and low-level) because it becomes an actionable plan.

I had a good bit of difficulty reviewing this document.  Right now it seems a bit confused in its presentation: parts of it are more like an outline or wishlist of features that don't exist yet (e.g. "look up wallet addresses staked into snode"), and other dive deep into the details (e.g. the "path streams" section) without enough description of the big picture of the feature to be able to fully understand and assess the details described.

I've left various suggestions below, as well as a handful of comments in the PR.

> @jagerman i fell down an analytical rabbit hole last week on this so i forget if this was meant to be a full on technical protocol spec or just a high level thing.

It needs to be both, but it *also* needs to be a proposal of how/why we are designing it the way it is.  (It's possible that this will eventually need to be split into multiple documents for different levels of understanding, but that is an editing task to be worried about (much) later in the process.)  More on this below.

# Suggestions

## Avoid scope creep

There's already too much scope creep in this document: for example, most of the "network data lookup stream" items are adding new features.  This PR is not the place for that.

That doesn't mean those features are not useful, it means those features are out of scope for the *initial* reimplementation of the network protocol.  The protocol needs to be able to be easily extensible to handle them, of course, and part of the protocol design needs to allow for protocol expansion, but for features that we don't have yet the protocol should only describe a generic request mechanism that is suitable for future request types.

## (Much) more elaboration

For the concrete requests that we do have, we need a concrete description of what the message looks like, all the data it contains, and so on.  For instance, what does a gossiped RC look like and what are the overall request, response, and error responses for updating one?  Or a path build request.  Or a snode lookup request.

## Avoid background knowledge

This document is currently heavily dependent on background knowledge of the existing protocol.  It should not be.  If pieces of the current protocol will be reused then write it out again, in full, because, unless there is a good reason why it must remain unchanged, it is just as much subject to critique as any *new* mechanisms being proposed here.  (And even if it must remain unchanged, it is still required here for this to be a complete reference of the protocol).

As I said above, we should aim to write this such that someone with *no* Lokinet knowledge can write a conforming client without looking at the current Lokinet implementation.

## What but also Why

The protocol design here needs more connection to high-level concepts so that this document can serve a purpose as not only the *what* of the protocol, but also the *why* of the protocol.  For instance, I have a lot of questions that I've left here in the path stream section, but I still don't have the big picture of *why* the stream state complexity of path streams is needed or what it allows that couldn't be done via another approach.

Abstractly:

1. We need to support feature X.
2. X requires an implementation of Y.
3. Here's Z.
4. Z is a proposed implementation of Y.

Reading this document right now I feel like I'm getting mostly just part 3, but understanding the protocol requires part 4, and proper discussion of the *design* of the protocol requires all of 2-4 (and maybe a bit of 1).

For a concrete example, see the pubkey list and pubkey list delta: in the current document there is no indication of what or who these are for or what problem having these endpoints solves.  I was going to comment on potential issues with download the entire list, or with the hash-diff mechanism, but then I realized my comments would only be based on my own guess of what these are here for.

## Review of current feature sets

I don't see anything about ONS lookups.

I don't know if there are other omissions, but would be worth going through again everything we need to support for *current* functionality to see if there are other pieces that are missing.
